### PR TITLE
Add Yiddish stemmer

### PR DESCRIPTION
**Pull request by urieli**
See https://github.com/snowballstem/snowball/pull/137

### CODE REVIEW

File: algorithms/yiddish/stemmer.tt [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  <li>Delete noun endings: יזם, יזמס.</li>
+  <li>Delete Hebraic plural ending: ים</li>
+  <li>Replace the Hebraic plural ending ות with ה.</li>
+  <li>Delete the diminutive endings: עלעך, עלע, לעך, עלעס, עלען.</li>
+  <li>Delete the noun endings: יסט, יסטן.</li>
+  <ul>
+    <li>Exceptions to the above: words ending in גיסט or שיסט.</li>
+  </ul>
+  <li>Delete the verb ending סטו.</li>
+  <li>Delete the uperlative endings: סטער, סטע, סטן.</li>
```
Comment (ojwb) [inline algorithms/yiddish/stemmer.tt:186]: I just noticed "uperlative" when checking the generated website - is that a typo for "superlative"?

**Comment by ojwb** [comment]
I think I've merged everything successfully, but if you spot any issues please report them.

I've also added Yiddish to the website demo at https://snowballstem.org/demo.html - it would be useful if you could quickly check that actually works properly.  I'm mostly wondering if anything extra needs doing for right-to-left alphabets (we already support Arabic there but I'm not sure if anybody actually checked that works properly either.)

**Comment by urieli** [comment]
The demo website https://snowballstem.org/demo.html works properly on both Firefox and Chrome. The only issue is that punctuation at the end of a paragraph isn't placed correctly. This can be fixed by adding the HTML attribute `dir="rtl"` to the `textarea` when the language is Yiddish or Arabic, or by adding the style `style="direction: rtl;"`. Is this page in github? If so, I could create a pull request to add the required attributes.

File: algorithms/yiddish/stemmer.tt (hunk 1 above)
Comment (urieli) [reply algorithms/yiddish/stemmer.tt:186]: Yes, definitely a typo for "superlative"! I'll add a new pull request for this, and also do a full reread of the text and code comments.

**Comment by urieli** [comment]
I found the demo page here: https://github.com/snowballstem/snowball-website/blob/master/demo.tt

However, I'm not familiar with `tt` files (Visual Studio Text Template?), so I may not be able to provide a PR to fix the minor issue in punctuation display.

**Comment by ojwb** [comment]
They're HTML templated with Perl Template Toolkit.  So it's just HTML, except for the `[%` ... `%]` directives.

Looking at demo.tt, we actually already have a special case to set RTL for Arabic (it was even me that added it in 71568748ca6ede3c328797d973e30517b71b1ded, but that was 4.5 years ago).

I've extended that to Yiddish in 8b4d55cfff49a40da92866bd9612e762a2c83851 (the website may take a few minutes to update, and you might need to force a reload with Shift-F5 or similar).  I'll note this in the contributing doc too.

Thanks for checking.

File: algorithms/yiddish/stemmer.tt (hunk 1 above)
Comment (ojwb) [reply algorithms/yiddish/stemmer.tt:186]: Thanks for confirming - no need for a PR, I've now pushed a fix.

**Comment by urieli** [comment]
Thanks Olly - the RTL works perfectly now for Yiddish!

**Comment by ojwb** [comment]
@urieli I think I've found a bug in the implementation:

```
      [substring] among(  
// [elidied other among cases]
        // Present participle ending -endik: delete if after a -ng, -nk, -n, -m, consonant+l, or vowel.    
        // Otherwise, delete just the -ndik part.  
        '{Ayen}{Nun}{Dalet}{Yud}{Kuf}'  
           ( R1  
             not (test('{Nun}{Giml}' or '{Nun}{Kuf}' or '{Nun}' or '{Mem}' or ('Lamed' consonant) or vowel) delete)    
             hop -5 '{Nun}{Dalet}{Yud}{Kuf}' delete  
           )  
      )  
```

The comment suggests that the `hop -5 '{Nun}{Dalet}{Yud}{Kuf}' delete` is intended to delete just the `{Nun}{Dalet}{Yud}{Kuf}` part, but `delete` deletes the slice, which is still set to `{Ayen}{Nun}{Dalet}{Yud}{Kuf}` by `[substring]`.  The algorithm description seems to confirm this (though I'm not familiar with the Hebrew alphabet so I may be looking at the wrong part).

Is my analysis right here?

If so, this could be fixed by adding `[` and `]` around the string in that line, but this line would actually be better written as:

```
<- '{Ayen}'
```

That would replace the slice (all 5 letters) with just `{Ayen}`.  This avoids the negative `hop`, which the language definition says causes `hop` to fail, but the current implementations actually move the cursor back and some of your code relies on this bug (which is why I was looking at the uses of negative `hop` in the first place).

Making my suggested change gives different output for some words in the test vocabulary.  If the change looks good to you, can you confirm I should just update the expected output for those words to match what the amended stemmer gives?

**Comment by ojwb** [comment]
@urieli I've also noticed a typo in the code quoted above - it uses `'Lamed'` not `'{Lamed}'`, so it checks the 5 Latin letters "Lamed" rather than the Hebrew character it should.

Changing that doesn't affect the results on the test vocabulary though.  It would be good to have test coverage for this case - what's an input where this `'{Lamed}' consonant` check makes a difference?

Also, the code to "skip verbal prefix" seems to have an extra check not in the written description of the algorithm on the website - the line after the "FIXME" comment I've added below:

```
        // If verbal prefix followed by Tsu- or Ge-, replace it
        (
// FIXME: Next line not in the description?
          test (('{Tsadek}{Vov}{Giml}{Nun}' or '{Tsadek}{Vov}{Kuf}{Tes}' or '{Tsadek}{Vov}{Kuf}{Nun}') not next)
          or                         
          ( ['{Giml}{Ayen}'] <- 'GE')   
          or                         
          (['{Tsadek}{Vov}'] <- 'TSU')
        )
```

The description of this part of the algorithm says "Prefixes include [...]" which could be read as suggesting that the written description doesn't contain a complete list.  If it doesn't, I think that needs fixing - someone should be able to implement the algorithm from scratch based on the description only.  If the wording is just imprecise, let me know and I'll adjust it to make it clear all the prefixes to check for are listed.

I also have some proposed changes to eliminate the other uses of negative hop, but these shouldn't affect the output of the algorithm (and don't at least for the test vocabulary) so I suggest we resolve the issues I've raised above first.

**Comment by urieli** [comment]
Hi Olly, thanks for these comments. I'll try to find time to review in the coming two weeks and let you know.

**Comment by ojwb** [comment]
Regarding the `Lamed` vs `{Lamed}` typo, I've realised the test vocabulary does actually cover this, but it's hidden without the -ndik fix (because the Ayen gets removed unconditionally currently, so the `Lamed` typo makes no difference).

So don't worry about coming up with a testcase for this.

**Comment by ojwb** [comment]
@urieli Did you get a chance to review yet?

If not are you likely to soon?

I'd like to make a new release soon.  It'd be good to resolve this first, though I could just disable the yiddish stemmer for now if it's going to take a while to sort this out.

**Comment by urieli** [comment]
@ojwb I'm sorry for not getting to this sooner. It's unlikely I'll have time this week, although I may have some time Friday afternoon. If not it will have to wait until next Wednesday. Do you have a release date in mind?

**Comment by ojwb** [comment]
Not a specific date.  It's been over a year since the last release which is really much too long, and last week I was asked about a new pypi upload.  So fairly soon would definitely be good.

Do let me know if there's any useful input I can provide.

**Pull request merged**